README
pypjlink is a python library and command line utility to interoperate with
projectors through pjlink protocol (http://pjlink.jbmia.or.jp/english/)

It is a fork of the work of Peter Ward (https://bitbucket.org/flowblok/pjlink).


Installation
============

Latest release can be found on pypi and installed with:
$ pip install


Command line usage
==================

Querying projector status is simple:
$ pjlink -p 10.1.1.1 power

and so it is sending commands:
$ pjlink -p 10.1.1.1 power off

A password will be requested if configured on the projector.

Please see help or manpage for complete usage options.


Python library usage
====================

The library supports the following python versions:
* python2.7
* python3.4
* python3.5

>>> from pypjlink import Projector
>>> projector = Projector.from_address('10.1.1.1')
>>> projector.get_power()
'off'
>>> projector.set_power('off')


To run tests, you can use tox to run against every supported python versions
$ tox

Or you can launch tests in current environment:

$ python -m unittest discover pypjlink.tests




CHANGELOG
==================
pypjlink changelog
==================


1.1.1 (2016-03-01)
==================

* added manual page
* password argument of authenticate() method can be a string, or omitted if
  the projector doesn't require authentication


1.1 (2016-02-02)
================

* Python 3 support


pypjlink/tests/utils.py
# -*- coding: utf-8 -*-


from io import StringIO, BytesIO

try:
    from unittest import mock
except ImportError:
    import mock



class MockProjSocket(object):
    """A context manager that mocks socket creation inside projector module
    Read from the mock returns response string passed in constructor.
    Write into the mock appends data in .written string attribute, besides
    calling .write mock method
    """

    mock_target = 'pypjlink.projector.socket.socket'

    def __init__(self, response):
        self._response = response
        self._mock_sock = None

    def __enter__(self):
        mock_socket_func = mock.patch(self.mock_target).__enter__()
        self._mock_sock = mock_socket_func.return_value
        stream = self._mock_sock.makefile.return_value
        import sys
        if sys.version_info.major == 2:
            buf = BytesIO(self._response.encode('utf-8'))
        else:
            buf = StringIO(self._response)

        stream.read = buf.read
        stream.written = ''

        def _write(data):
            stream.written += data
        stream.write.side_effect = _write

        return stream

    def __exit__ (self, exc_type, exc_value, traceback):
        self._mock_sock.__exit__()


pypjlink/tests/test_projector.py
# -*- coding: utf-8 -*-

from __future__ import unicode_literals

import hashlib
import unittest

from pypjlink.projector import Projector, POWER_STATES
from .utils import MockProjSocket


NO_AUTH_RESPONSE = 'PJLINK 0\r'


class AuthenticationTC(unittest.TestCase):
    def test_no_auth(self):
        """test detection of no auth projector"""
        power_response = '%1POWR=1\r'
        with MockProjSocket(NO_AUTH_RESPONSE + power_response) as mock_stream:

            proj = Projector.from_address('projhost')
            self.assertFalse(proj.authenticate(lambda: ''))
            # since projector said no auth required, I shouldn't have written
            # anything to it
            self.assertFalse(mock_stream.write.called)

    def test_auth(self):
        """test authentication"""
        auth_response = 'PJLINK 1 00112233\r'
        power_response = '%1POWR=1\r'

        with MockProjSocket(auth_response + power_response) as mock_stream:
            proj = Projector.from_address('projhost')
            self.assertTrue(proj.authenticate(lambda: 'p'))
            # test write of authentication
            self.assertEqual(
                mock_stream.written,
                self._md5_auth_code('00112233', 'p') + '%1POWR ?\r')

    def test_wrong_auth(self):
        """test failed authentication"""
        response = (
            'PJLINK 1 00112233\r'
            'PJLINK ERRA\r'
        )

        with MockProjSocket(response) as mock_stream:
            proj = Projector.from_address('projhost')
            self.assertFalse(proj.authenticate(lambda: 'p'))
            self.assertEqual(
                mock_stream.written,
                self._md5_auth_code('00112233', 'p') + '%1POWR ?\r')

    def test_string_password(self):
        """since 1.1.1, password can be a string in authenticate() method"""
        auth_response = 'PJLINK 1 00112244\r'
        power_response = '%1POWR=1\r'
        with MockProjSocket(auth_response + power_response) as mock_stream:
            proj = Projector.from_address('projhost')
            self.assertTrue(proj.authenticate('ps'))
            # test write of authentication
            self.assertEqual(
                mock_stream.written,
                self._md5_auth_code('00112244', 'ps') + '%1POWR ?\r')

    def test_no_password_no_auth(self):
        """since 1.1.1, password can be omitted from authenticate() method
        if projector doesn't need a password"""
        power_response = '%1POWR=1\r'
        with MockProjSocket(NO_AUTH_RESPONSE + power_response) as mock_stream:
            proj = Projector.from_address('projhost')
            self.assertFalse(proj.authenticate())
            self.assertFalse(mock_stream.write.called)

    def test_no_password_auth_required(self):
        """if projector needs a password but is missing in authenticate() there
        should be an error"""
        auth_response = 'PJLINK 1 00112255\r'
        power_response = '%1POWR=1\r'
        with MockProjSocket(auth_response + power_response) as mock_stream:
            proj = Projector.from_address('projhost')
            with self.assertRaises(RuntimeError):
                proj.authenticate()
            self.assertFalse(mock_stream.write.called)

    def _md5_auth_code(self, salt, password):
        return hashlib.md5((salt + password).encode('utf-8')).hexdigest()


class NameTC(unittest.TestCase):
    def test_unicode(self):
        """test utf-8 chars in projector name"""
        response = NO_AUTH_RESPONSE + '%1NAME=à€\r'
        with MockProjSocket(response) as mock_stream:
             proj = Projector.from_address('projhost')
             proj.authenticate(lambda: '')
             name = proj.get_name()
             self.assertEqual(name, 'à€')


class PowerTC(unittest.TestCase):
    def test_get(self):
        response = NO_AUTH_RESPONSE + '%1POWR={}\r'.format(
            POWER_STATES['cooling'])
        with MockProjSocket(response) as mock_stream:
            proj = Projector.from_address('projhost')
            proj.authenticate(lambda: '')
            self.assertEqual(proj.get_power(), 'cooling')

    def test_set(self):
        response = NO_AUTH_RESPONSE + '%1POWR=OK\r'
        with MockProjSocket(response) as mock_stream:
            proj = Projector.from_address('projhost')
            proj.authenticate(lambda: '')
            proj.set_power('off')

            self.assertEqual(
                mock_stream.written, '%1POWR {}\r'.format(POWER_STATES['off']))


pypjlink/protocol.py
# -*- coding: utf-8 -*-

import sys


def read_until(f, term):
    data = []
    c = f.read(1)
    while c != term:
        data.append(c)
        c = f.read(1)
    data = ''.join(data)
    if sys.version_info.major == 2:
        data = data.decode('utf-8')
    return data

def to_binary(body, param, sep=' '):
    assert body.isupper()

    assert len(body) == 4
    assert len(param) <= 128

    return '%1' + body + sep + param + '\r'

def parse_response(f, data=''):
    if len(data) < 7:
        data += read(f, 2 + 4 + 1 - len(data))

    header = data[0]
    assert header == '%'

    version = data[1]
    # only class 1 is currently defined
    assert version == '1'

    body = data[2:6]
    # commands are case-insensitive, but let's turn them upper case anyway
    # this will avoid the rest of our code from making this mistake
    # FIXME: AFAIR this takes the current locale into consideration, it shouldn't.
    body = body.upper()

    sep = data[6]
    assert sep == '='

    param = read_until(f, '\r')

    return (body, param)

# python 3 socket makefile is already unicode in text mode, i do the same on
# python 2
if sys.version_info.major == 2:
    def read(f, n):
        return f.read(n).decode('utf-8')
else:
    def read(f, n):
        return f.read(n)


ERRORS = {
    'ERR1': 'undefined command',
    'ERR2': 'out of parameter',
    'ERR3': 'unavailable time',
    'ERR4': 'projector failure',
}

def send_command(f, req_body, req_param):
    data = to_binary(req_body, req_param)
    f.write(data)
    f.flush()

    resp_body, resp_param = parse_response(f)
    assert resp_body == req_body

    if resp_param in ERRORS:
        return False, ERRORS[resp_param]
    return True, resp_param


pypjlink/projector.py
# -*- coding: utf-8 -*-

from __future__ import unicode_literals

import hashlib
import socket
import sys

from pypjlink import protocol

class ProjectorError(Exception):
    pass

reverse_dict = lambda d: dict(zip(d.values(), d.keys()))

POWER_STATES = {
    'off': '0',
    'on': '1',
    'cooling': '2',
    'warm-up': '3',
}
POWER_STATES_REV = reverse_dict(POWER_STATES)

SOURCE_TYPES = {
    'RGB': '1',
    'VIDEO': '2',
    'DIGITAL': '3',
    'STORAGE': '4',
    'NETWORK': '5',
}
SOURCE_TYPES_REV = reverse_dict(SOURCE_TYPES)

MUTE_VIDEO = 1
MUTE_AUDIO = 2
MUTE_STATES_REV = {
    '11': (True, False),
    '21': (False, True),
    '31': (True, True),
    '30': (False, False),
}

ERROR_STATES_REV = {
    '0': 'ok',
    '1': 'warning',
    '2': 'error',
}

class Projector(object):
    def __init__(self, f):
        self.f = f

    @classmethod
    def from_address(cls, address, port=4352):
        """build a Projector from a ip address"""
        sock = socket.socket()
        sock.connect((address, port))
        # in python 3 I need to specify newline, otherwise read hangs
        # in "PJLINK 0\r"
        # I expect socket file to return byte strings in python 2 and
        # unicode strings in python 3
        if sys.version_info.major == 2:
            f = sock.makefile()
        else:
            f = sock.makefile(mode='rw', newline='\r')
        return cls(f)

    def authenticate(self, password=None):
        # I'm just implementing the authentication scheme designed in the
        # protocol. Don't take this as any kind of assurance that it's secure.

        data = protocol.read(self.f, 9)
        assert data[:7] == 'PJLINK '
        security = data[7]
        if security == '0':
            return None
        data += protocol.read(self.f, 9)
        assert security == '1'
        assert data[8] == ' '
        salt = data[9:17]
        assert data[17] == '\r'

        if password is None:
            raise RuntimeError('projector needs a password')

        if callable(password):
            password = password()

        pass_data = (salt + password).encode('utf-8')
        pass_data_md5 = hashlib.md5(pass_data).hexdigest()

        # we *must* send a command to complete the procedure,
        # so we just get the power state.

        cmd_data = protocol.to_binary('POWR', '?')
        self.f.write(pass_data_md5 + cmd_data)
        self.f.flush()

        # read the response, see if it's a failed auth
        data = protocol.read(self.f, 7)
        if data == 'PJLINK ':
            # should be a failed auth if we get that
            data += protocol.read(self.f, 5)
            assert data == 'PJLINK ERRA\r'
            # it definitely is
            return False

        # good auth, so we should get a reply to the command we sent
        body, param = protocol.parse_response(self.f, data)

        # make sure we got a sensible response back
        assert body == 'POWR'
        if param in protocol.ERRORS:
            raise ProjectorError(protocol.ERRORS[param])

        # but we don't care about the value if we did
        return True

    def get(self, body):
        success, response = protocol.send_command(self.f, body, '?')
        if not success:
            raise ProjectorError(response)
        return response

    def set(self, body, param):
        success, response = protocol.send_command(self.f, body, param)
        if not success:
            raise ProjectorError(response)
        assert response == 'OK'

    # Power

    def get_power(self):
        param = self.get('POWR')
        return POWER_STATES_REV[param]

    def set_power(self, status, force=False):
        if not force:
            assert status in ('off', 'on')
        self.set('POWR', POWER_STATES[status])

    # Input

    def get_input(self):
        param = self.get('INPT')
        source, number = param
        source = SOURCE_TYPES_REV[source]
        number = int(number)
        return (source, number)

    def set_input(self, source, number):
        source = SOURCE_TYPES[source]
        number = str(number)
        assert number in '123456789'
        self.set('INPT', source + number)

    # A/V mute

    def get_mute(self):
        param = self.get('AVMT')
        return MUTE_STATES_REV[param]

    def set_mute(self, what, state):
        assert what in (MUTE_VIDEO, MUTE_AUDIO, MUTE_VIDEO | MUTE_AUDIO)
        what = str(what)
        assert what in '123'
        state = '1' if state else '0'
        self.set('AVMT', what + state)

    # Errors

    def get_errors(self):
        param = self.get('ERST')
        errors = 'fan lamp temperature cover filter other'.split()
        assert len(param) == len(errors)
        return dict((key, ERROR_STATES_REV[value]) for key, value in zip(errors, param))

    # Lamps

    def get_lamps(self):
        param = self.get('LAMP')
        assert len(param) <= 65
        values = param.split(' ')
        assert len(values) <= 16 and len(values) % 2 == 0

        lamps = []
        for time, state in zip(values[::2], values[1::2]):
            time = int(time)
            state = bool(int(state))
            lamps.append((time, state))

        assert len(lamps) <= 8
        return lamps

    # Input list

    def get_inputs(self):
        param = self.get('INST')
        assert len(param) <= 95

        values = param.split(' ')
        assert len(values) <= 50

        inputs = []
        for value in values:
            source, number = value
            source = SOURCE_TYPES_REV[source]
            assert number in '123456789'
            number = int(number)
            inputs.append((source, number))

        return inputs

    # Projector info

    def get_name(self):
        param = self.get('NAME')
        assert len(param) <= 64
        return param

    def get_manufacturer(self):
        param = self.get('INF1')
        assert len(param) <= 32
        # stupidly, this is not defined as utf-8 in the spec. :(
        return param

    def get_product_name(self):
        param = self.get('INF2')
        assert len(param) <= 32
        # stupidly, this is not defined as utf-8 in the spec. :(
        return param

    def get_other_info(self):
        param = self.get('INFO')
        assert len(param) <= 32
        return param

    # TODO: def get_class(self): self.get('CLSS')
    # once we know that class 2 is, and how to deal with it


pypjlink/cli.py
import argparse

try:
    from ConfigParser import (
        NoSectionError,
        SafeConfigParser as ConfigParser
    )
except ImportError:
    from configparser import (
        NoSectionError,
        SafeConfigParser as ConfigParser
    )

from getpass import getpass
from os import path
import sys

import appdirs

from pypjlink import Projector
from pypjlink import projector
from pypjlink.cliutils import make_command, print_error

def cmd_power(p, state=None):
    if state is None:
        print(p.get_power())
    else:
        p.set_power(state)

def cmd_input(p, source, number):
    if source is None:
        source, number = p.get_input()
        print(source, number)
    else:
        p.set_input(source, number)

def cmd_inputs(p):
    for source, number in p.get_inputs():
        print('%s-%s' % (source, number))

def cmd_mute_state(p):
    video, audio = p.get_mute()
    print('video:', 'muted' if video else 'unmuted')
    print('audio:', 'muted' if audio else 'unmuted')

def cmd_mute(p, what):
    if what is None:
        return cmd_mute_state(p)
    what = {
        'video': projector.MUTE_VIDEO,
        'audio': projector.MUTE_AUDIO,
        'all': projector.MUTE_VIDEO | projector.MUTE_AUDIO,
    }[what]
    p.set_mute(what, True)

def cmd_unmute(p, what):
    if what is None:
        return cmd_mute_state(p)
    what = {
        'video': projector.MUTE_VIDEO,
        'audio': projector.MUTE_AUDIO,
        'all': projector.MUTE_VIDEO | projector.MUTE_AUDIO,
    }[what]
    p.set_mute(what, False)

def cmd_info(p):
    info = [
        ('Name', p.get_name()),
        ('Manufacturer', p.get_manufacturer()),
        ('Product Name', p.get_product_name()),
        ('Other Info', p.get_other_info())
    ]
    for key, value in info:
        print('%s: %s' % (key, value))

def cmd_lamps(p):
    for i, (time, state) in enumerate(p.get_lamps(), 1):
        print('Lamp %d: %s (%d hours)' % (
            i,
            'on' if state else 'off',
            time,
        ))

def cmd_errors(p):
    for what, state in p.get_errors().items():
        print('%s: %s' % (what, state))

def make_parser():
    parser = argparse.ArgumentParser()
    parser.add_argument('-p', '--projector')

    sub = parser.add_subparsers(dest='command', title='command')
    sub.required = True

    power = make_command(sub, 'power', cmd_power)
    power.add_argument('state', nargs='?', choices=('on', 'off'))

    inpt = make_command(sub, 'input', cmd_input)
    inpt.add_argument('source', nargs='?', choices=projector.SOURCE_TYPES)
    inpt.add_argument('number', nargs='?', choices='123456789', default='1')

    make_command(sub, 'inputs', cmd_inputs)

    mute = make_command(sub, 'mute', cmd_mute)
    mute.add_argument('what', nargs='?', choices=('video', 'audio', 'all'))

    unmute = make_command(sub, 'unmute', cmd_unmute)
    unmute.add_argument('what', nargs='?', choices=('video', 'audio', 'all'))

    make_command(sub, 'info', cmd_info)
    make_command(sub, 'lamps', cmd_lamps)
    make_command(sub, 'errors', cmd_errors)

    return parser

def resolve_projector(projector):
    password = None

    # host:port
    if projector is not None and ':' in projector:
        host, port = projector.rsplit(':', 1)
        port = int(port)

    # maybe defined in config
    else:
        appdir = appdirs.user_data_dir('pjlink')
        conf_file = path.join(appdir, 'pjlink.conf')

        try:
            config = ConfigParser({'port': '4352', 'password': ''})
            with open(conf_file, 'r') as f:
                config.readfp(f)

            section = projector
            if projector is None:
                section = 'default'

            host = config.get(section, 'host')
            port = config.getint(section, 'port')
            password = config.get(section, 'password') or None

        except (NoSectionError, IOError):
            if projector is None:
                raise KeyError('No default projector defined in %s' % conf_file)

            # no config file, or no projector defined for this host
            # thus, treat the projector as a hostname w/o port
            host = projector
            port = 4352


    return host, port, password

def main():
    parser = make_parser()
    args = parser.parse_args()

    kwargs = dict(args._get_kwargs())
    func = kwargs.pop('__func__')
    kwargs.pop('command', None)

    projector = kwargs.pop('projector')
    host, port, password = resolve_projector(projector)

    if not password:
        password = getpass

    proj = Projector.from_address(host, port)
    rv = proj.authenticate(password)
    if rv is False:
        print_error('Incorrect password.')
        return

    func(proj, **kwargs)

if __name__ == '__main__':
    main()
